[SUREFIRE-1827] The console output is not flushed

Tibor17 opened the pull request.

Tibor17 commented:
Missing the unit tests for this functionality.
I will work on it tomorrow.

rmannibucau commented:
Can be good to at least support a system property for flush interval and not leak the flush thread (stop the scheduler properly) but otherwise looks good.

Tibor17 commented:
@rmannibucau 
I have added two unit tests. But the important thing is the manual test i made. I have expected that you would like to make the same test on your PC. So i changed my commit to 100 millis from previous 200 millis because 100 millis makes the visual effect smooth. So i do not see the business value of config paramter. I would not bother the user with more and more config params because anyway they will be lost. Better if it just works with ideally no new config params. This was my test:

        for ( int i = 0; i < 50; i++ )
```
        {
            System.out.println( "01234567890123456789012345678901234567890123456789" );
            Thread.sleep(10);
        }
```

olamy commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/LegacyMasterProcessChannelProcessorFactory.java:
```diff
@@ -36,8 +36,10 @@
-    implements MasterProcessChannelProcessorFactory
+    extends AbstractMasterProcessChannelProcessorFactory
 {
+    private static final int FLUSH_PERIOD_MILLIS = 100;
```
would be good to have this configurable. At least Integer.getInteger(""whatever name you like", 100)

Tibor17 replied:
But it does not make sense. It is not a business feature and it is hidden for the user. The only purpose is that it just works, prints the console, nothing special.

rmannibucau replied:
Fact is it depends the OS and potentially the tests so having a way to change a good default (>= 150ms IMHO) is good.

eolivelli replied:
I can see a use case of having in CI this value to be set to "infinite" in order to save resources.
But adding a new knob means that we will have to maintain it.
I am leaning toward keeping an hardcoded constant, but I don't feel so strong about it.

eolivelli commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/AbstractMasterProcessChannelProcessorFactory.java:
```diff
@@ -0,0 +1,106 @@
+                {
+                    try
+                    {
+                        channel.write( ByteBuffer.allocate( 0 ) );
```
we can create a constant and save this buffer allocation

Tibor17 replied:
sure, i will

rmannibucau commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/AbstractMasterProcessChannelProcessorFactory.java:
```diff
@@ -0,0 +1,106 @@
+
+    public AbstractMasterProcessChannelProcessorFactory()
+    {
+        flusher = newScheduledThreadPool( 1, newDaemonThreadFactory( STREAM_FLUSHER ) );
```
Shouldnt be daemon, there is no reason and can only hide bugs.

Tibor17 replied:
The daemon will be safely stopped when we call `System.exit()` in `ForkedBooter`.
This principle was used even before i entered the project.
It will guarantee that the thread would not prolong the execution time.
We were working with Jonathan Bell on fast JVM startup and teardown and we made the teardown from 200 minnils to 40.

rmannibucau replied:
I think there is never any reason to not handle the lifecycle of the object you create properly.
System.exit cleanup order is not deterministic (in particular across JVM and versions) and you literally let a thread leak there assuming the OS cleans it up.
It also depends your glic version (see https://www.man7.org/linux/man-pages/man2/_exit.2.html).
So even not technically I would shutdown properly any started thread in the app + here you will not add any latency since it will exit immediately using shutdownnow + awaitTermination (will just do a state check so will be way faster than 100ms).

eolivelli commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/AbstractMasterProcessChannelProcessorFactory.java:
```diff
@@ -0,0 +1,106 @@
+    {
+        try
+        {
+            doPrivileged( new PrivilegedAction<Object>()
```
just  for curiosity, why are you using "doPrivileged" ?

Tibor17 replied:
See the junit3 ITs. They are with the SecurityManager.
You always have to do it, even in the ForkedBooter we have priviledged call.

rmannibucau replied:
If you care about fine level perf you can likely use the standard pattern (if (sm == null) doShutdown() else doPriviledgedShutdown()). Not sure the 3-4 allocations change much things in this particular case but it is always a bit surprising to see a doPriviledged not conditionned (guess it became a so common pattern than we think less about it now?).

eolivelli commented on surefire-booter/src/main/java/org/apache/maven/surefire/booter/spi/AbstractMasterProcessChannelProcessorFactory.java:
```diff
@@ -0,0 +1,106 @@
+                              @Override
+                              public Object run()
+                              {
+                                  flusher.shutdown();
```
+1

rmannibucau replied:
Maybe call shutdownNow at least and cancel the scheduked task.
Then calling awaitTermination should be immediate in this case which avoids side effects of this method even if unlikely (= makes it future proof)

jglick commented:
Amending #240 IIUC.

jira-importer commented:
Resolve #3017
